### PR TITLE
[WIP]exec: supprt exec as a specific user

### LINKED ISSUE #30656: Support the user flag from docker exec in kubectl exec
It looks like `docker exec` is being used as the backend for `kubectl exec`. `docker exec` has the `--user` flag, which allows you to run a command as a particular user.  This same functionality doesn't exist in Kubernetes.  

Our use case is that we spin up pods, and execute untrusted code in them.  However, there are times when after creating the pod, we need to run programs that need root access (they need to access privileged ports, etc).

We don't want to run the untrusted code as root in the container, which prevents us from just escalating permissions for all programs.

I looked around for references to this problem, but only found this StackOverflow answer from last year -- http://stackoverflow.com/questions/33293265/execute-command-into-kubernetes-pod-as-other-user .

There are some workarounds to this, such as setting up a server in the container that takes commands in, or defaulting to root, but dropping to another user before running untrusted code.  However, these workarounds break nice Kubernetes/Docker abstractions and introduce security holes.

### PR DESCRIPTION
**Pull request by max88991**
Signed-off-by: Lei Gong <lgong@alauda.io>

/kind feature

**What this PR does / why we need it**:

Supprt exec as a specific user

**Which issue(s) this PR fixes**:

Fixes #30656

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Add a `-u` flag for `kubectl exec`
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

**Comment by smarterclayton** [comment]
Because this has security implications, it’ll need to be discussed in sig-auth.  Currently a container always runs as the user identified in the pod spec or image, so an authorization check performed at creation time is sufficient to ensure a pod has a specific security profile.  Allowing user to be specified at exec time would require some level of verification that a non root container cannot run as root, as well as ensuring existing admission control checks cover this new path.

Please email the kubernetes-sig-auth google group to get this PR on their agenda to discuss.

**Comment by mattjmcnaughton** [comment]
> Because this has security implications, it’ll need to be discussed in sig-auth. Currently a container always runs as the user identified in the pod spec or image, so an authorization check performed at creation time is sufficient to ensure a pod has a specific security profile. Allowing user to be specified at exec time would require some level of verification that a non root container cannot run as root, as well as ensuring existing admission control checks cover this new path.
> 
> Please email the kubernetes-sig-auth google group to get this PR on their agenda to discuss.

+1

This seems like a substantial change for which we'd want to have a [Kubernetes Enhancement Proposal](https://github.com/kubernetes/enhancements/tree/master/keps) - cc @davidxia who mentioned they were planning to write one.

The code will be useful to have for the KEP discussion, and hopefully provide the starting point for the feature should the KEP be approved. Could I trouble you to prefix your PR title with `WIP` in the meantime? Thanks!

File: staging/src/k8s.io/kubectl/pkg/cmd/exec/exec.go [hunk 1]
```diff
@@ -350,6 +352,7 @@ func (p *ExecOptions) Run() error {
 		req.VersionedParams(&corev1.PodExecOptions{
 			Container: containerName,
 			Command:   p.Command,
+			User:      p.Username,
```
[the remark this answers is not in the document]
Comment (max88991) [reply staging/src/k8s.io/kubectl/pkg/cmd/exec/exec.go]: Yes, check added

**Comment by max88991** [comment]
ping @ mattjmcnaughton @smarterclayton please take a took at the KEP:
https://github.com/kubernetes/enhancements/pull/1224

**Comment by josiahbjorgaard** [comment]
Bug triage for 1.17 here. Code freeze is Nov. 18th. Will this PR be merged by then? If not, can we move it out of the milestone?

**Comment by josiahbjorgaard** [comment]
@dims Are you still intending this for v1.17?

**Comment by ahmetb** [comment]
FWIW krew has a kubectl plugin that seems to be doing what this does:

https://github.com/jordanwilson230/kubectl-plugins/blob/master/kubectl-ssh
https://github.com/jordanwilson230/kubectl-plugins/tree/krew#kubectl-exec-as

```
$ kubectl krew info exec-as
NAME: exec-as
URI: https://github.com/jordanwilson230/kubectl-plugins/archive/v1.0.2-krew.zip
SHA256: 9bd734b1dc08fc59df3f40418e92b228a830098a27257d454deccc5aa1384cf4
VERSION: v1.0.2-krew
HOMEPAGE: https://github.com/jordanwilson230/kubectl-plugins/tree/krew#kubectl-exec-as
DESCRIPTION:
Run `kubectl exec-as` for usage. Example, `kubectl exec-as -u root zookeeper-0`.
CAVEATS:
\
 |  The node which the pod is running on cannot have more than one taint.
/
```

**Pull request closed without merging**